### PR TITLE
OSDOCS-11180: creating workload partitioning section

### PR DESCRIPTION
**Pull request by aravipra**
Version(s):
4.17+

Issue:
[OSDOCS-11180](https://issues.redhat.com/browse/OSDOCS-11180)

Link to docs preview:
[workload partitioning](https://80198--ocpdocs-pr.netlify.app/microshift/latest/microshift_configuring/microshift-workload-partitioning.html)

QE review:
- [x] QE has approved this change.

SME review:
- [x] SME has approved this change.

### CODE REVIEW

File: microshift_configuring/microshift-workload-partitioning.adoc [hunk 1]
```diff
@@ -0,0 +1,13 @@
+:_mod-docs-content-type: ASSEMBLY
+[id="microshift-workload-partitioning"]
+= Workload partitioning
+include::_attributes/attributes-microshift.adoc[]
+:context: microshift-workload-partitioning
+
+toc::[]
+
+Workload partitioning separates compute node CPU resources into distinct CPU sets. The primary objective is to limit the amount of CPU usage for all control plane components which reserves rest of the device CPU resources for workloads of the user.
```
Comment (eslutsky) [inline microshift_configuring/microshift-workload-partitioning.adoc:9]: `compute node CPU`, in Microshift we don't have multiple nodes, we have one node which also acts as compute , the rest of the  the goal is correct .

File: modules/microshift-enabling-workload-partitioning.adoc [hunk 2]
```diff
@@ -0,0 +1,106 @@
+// Module included in the following assemblies:
+//
+// * microshift_configuring/microshift-workload-partitioning.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="microshift-enabling-workload-partitioning_{context}"]
+= Enabling workload partitioning
+
+System configuration changes has to be done to enable workload partitioning on {microshift-short}. This can be done by making the following changes:
+* Add Kubelet configuration to enable and configure CPU Manager for the workloads.
```
Comment (eslutsky) [inline modules/microshift-enabling-workload-partitioning.adoc:10]: in the preview page this doesn't appear at the bullets.
see:
image

File: modules/microshift-enabling-workload-partitioning.adoc [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+# ...
+kubelet:
+  reservedSystemCPUs: 0,6,7<1>
+  cpuManagerPolicy: static
+  cpuManagerPolicyOptions:
+    full-pcpus-only: "true"<2>
+  cpuManagerReconcilePeriod: 5s
+# ...
+----
+<1> exclusive cpuset for the system/microshift daemons and the interrupts/timers.
```
Comment (eslutsky) [inline modules/microshift-enabling-workload-partitioning.adoc:45]: this  for  kubelet system daemon its not directly affecting Microshift daemon. we have another configurations for Microshift (systemd  unit configuration).

File: microshift_configuring/microshift-workload-partitioning.adoc (hunk 1 above)
Comment (aravipra) [reply microshift_configuring/microshift-workload-partitioning.adoc:9]: done

File: modules/microshift-enabling-workload-partitioning.adoc (hunk 2 above)
Comment (aravipra) [reply modules/microshift-enabling-workload-partitioning.adoc]: this sentence is updated

File: modules/microshift-enabling-workload-partitioning.adoc (hunk 3 above)
Comment (aravipra) [reply modules/microshift-enabling-workload-partitioning.adoc:45]: done

File: modules/microshift-enabling-workload-partitioning.adoc [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  cpuManagerPolicyOptions:
+    full-pcpus-only: "true"<2>
+  cpuManagerReconcilePeriod: 5s
+# ...
+----
+<1> exclusive cpuset for the system daemons and the interrupts/timers.
+<2> kubelet configuration sets the `CPUManager policy` option to `full-pcpus-only` to ensure allocation of whole cores to a containers workload.
+
+. CRI-O configuration:
+* Update the CRI-O configuration file.
```
Comment (eslutsky) [inline modules/microshift-enabling-workload-partitioning.adoc:49]: its missing  the CRIO configuration file path, it should override the default configuration that already exists  by default `11-microshift-ovn.conf` so the config can be for example:  `/etc/crio/crio.conf.d/20-microshift-workload-partition.conf`  .

Comment (aravipra) [reply modules/microshift-enabling-workload-partitioning.adoc:49]: I have made changes. let me know if the changes are correct.

I have few questions:
1. Is there any links or references to be provided for more information . eg:  building blueprint?
2. Should the example blue print be added? examples are already provided in sections in the current doc.

Comment (eslutsky) [reply modules/microshift-enabling-workload-partitioning.adoc]: yes the changes with the crio configuration path looks correct, 
1. most of the example for blueprints are using `[[customizations.files]]` which can be found [here](https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/composing_a_customized_rhel_system_image/creating-system-images-with-composer-command-line-interface_composing-a-customized-rhel-system-image?extIdCarryOver=true&sc_cid=701f2000001Css5AAC#specifying_customized_files_in_the_blueprint),
2. I dont think you need to provide the sample example separately in blueprint format.

**Comment by dhensel-rh** [comment]
Is there an assumption that the VM is deployed with a certain number of cores ?

**Comment by aravipra** [comment]
> Is there an assumption that the VM is deployed with a certain number of cores ?

@eslutsky please can you confirm?

**Comment by eslutsky** [comment]
@ashiw

> > Is there an assumption that the VM is deployed with a certain number of cores ?
> 
> the numbers `0,6,7` is an example when we want to allocated 3 CPUs for the control plane, in case we have a machine with a total of 8 (0-7)
@aravipra , it does make sense to mention  it, because we specify that CPUset  (0,6,7) through the docs.

File: modules/microshift-enabling-workload-partitioning.adoc [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+//
+// * microshift_configuring/microshift-workload-partitioning.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="microshift-enabling-workload-partitioning_{context}"]
+= Enabling workload partitioning
+
+To enable workload partitioning on {microshift-short}, make the following configuration changes:
+
+* Update the {microshift-short} `config.yaml` file to include the Kubelet configuration file.
```
Comment (GroceryBoyJr) [inline modules/microshift-enabling-workload-partitioning.adoc:11]:
```suggestion
* Update the {microshift-short} `config.yaml` file to include the kubelet configuration file.
```

File: modules/microshift-enabling-workload-partitioning.adoc [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+* Create the CRI-O systemd and configuration files.
+* Create and update the systemd configuration file for the {microshift-short} and CRI-O services respectively.
+
+.Procedure
+
+. Update the {microshift-short} `config.yaml` file to include the Kubelet configuration file in order to enable and configure CPU Manager for the workloads:
```
Comment (GroceryBoyJr) [inline modules/microshift-enabling-workload-partitioning.adoc:17]:
```suggestion
. Update the {microshift-short} `config.yaml` file to include the kubelet configuration file in order to enable and configure CPU Manager for the workloads:
```

File: modules/microshift-enabling-workload-partitioning.adoc [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+* Create the Kubelet configuration file in the path `/etc/kubernetes/openshift-workload-pinning`. The Kubelet configuration directs the kubelet to modify the node resources based on the capacity and allocatable CPUs.
```
Comment (GroceryBoyJr) [inline modules/microshift-enabling-workload-partitioning.adoc:18]:
```suggestion
* Create the kubelet configuration file in the path `/etc/kubernetes/openshift-workload-pinning`. The kubelet configuration directs the kubelet to modify the node resources based on the capacity and allocatable CPUs.
```

File: modules/microshift-enabling-workload-partitioning.adoc [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
++
+.Kubelet configuration example
```
Comment (GroceryBoyJr) [inline modules/microshift-enabling-workload-partitioning.adoc:20]:
```suggestion
.kubelet configuration example
```

File: modules/microshift-enabling-workload-partitioning.adoc [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+kubelet:
+  reservedSystemCPUs: 0,6,7 <1>
+  cpuManagerPolicy: static
+  cpuManagerPolicyOptions:
+    full-pcpus-only: "true" <2>
+  cpuManagerReconcilePeriod: 5s
+# ...
+----
+<1> Exclusive cpuset for the system daemons and the interrupts/timers.
+<2> Kubelet configuration sets the `CPUManagerPolicyOptions` option to `full-pcpus-only` to ensure allocation of whole cores to the containers workload.
```
Comment (GroceryBoyJr) [inline modules/microshift-enabling-workload-partitioning.adoc:47]:
```suggestion
<2> The kubelet configuration sets the `CPUManagerPolicyOptions` option to `full-pcpus-only` to ensure allocation of whole cores to the containers workload.
```

**Review by GroceryBoyJr — COMMENTED** [review]
LGTM, only one small nit: "kublet" is lowercase, not capital "K", per the style guide: https://redhat-documentation.github.io/supplementary-style-guide/#kubelet

**Review by GroceryBoyJr — COMMENTED** [review]
Also, style guide asks us to use "to" instead of "in order to".

File: modules/microshift-enabling-workload-partitioning.adoc (hunk 5 above)
Comment (aravipra) [reply modules/microshift-enabling-workload-partitioning.adoc:11]: done

File: modules/microshift-enabling-workload-partitioning.adoc (hunk 6 above)
Comment (aravipra) [reply modules/microshift-enabling-workload-partitioning.adoc:17]: done

File: modules/microshift-enabling-workload-partitioning.adoc (hunk 7 above)
Comment (aravipra) [reply modules/microshift-enabling-workload-partitioning.adoc:18]: done

File: modules/microshift-enabling-workload-partitioning.adoc (hunk 8 above)
Comment (aravipra) [reply modules/microshift-enabling-workload-partitioning.adoc:20]: done

File: modules/microshift-enabling-workload-partitioning.adoc (hunk 9 above)
Comment (aravipra) [reply modules/microshift-enabling-workload-partitioning.adoc:47]: done

**Pull request merged**